### ISSUE TITLE
Improve Dagger Logging by ensuring we catch early exits, upload on early exits and log to appropriate places

### ISSUE THREAD

**Issue by bnchrch** [airbytehq/airbyte#28423]
# Overview
There a few issue we can address with the current dagger logs

# Issues

## Dagger logs are hard to parse which line relates to which step
The log lines for all pipelines and steps are output in a fashion that mixes them together

**Possible Solution(s)**
Dagger has a concept of “named sub-pipeline” that we should use more deeply to isolate steps in the Dagger logs. 
But we’d have to re-architecture the code a bit:
- The pipeline object is built from a dagger client or container and return a client or a container
- We should declare a Step.dagger_client attribute and seed if with context.dagger_client.pipeline(step_name)
- Environment.py functions should take Step as an input and not context object. To access the dagger client we’d call step.dagger_client instead of context.dagger_client 

## Dagger logs are hard to find
During a local error there is no output directing you to the dagger logs
On an early exit there may be no log file at all
**Possible Solution(s)**
1. When finishing a local run, output the location of the logs
2. Output each log section to a separate file
2. Set –show-dagger-logs to true by default for local runs
3.  Use Dagger’s experimental TUI by wrapping airbyte-ci call with dagger run: dagger run airbyte-ci connectors –name=source-pokeapi test
4.  

# Solution Criteria
1. On a hard error, we still capture dagger logs (likely the sys.exit(1) [here](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/utils.py#L468) is causing us to loose logs)
2. Invert `--dagger-logs` so that this is on by default and disable them in CI
3. Ensure we upload to GCS on early exit

**Comment by bnchrch** [comment]
Grooming:
- Is it possible to use Client.log_output to log to stderr AND a file? https://github.com/airbytehq/airbyte/blob/bnchrch/ci/poetry-pipelines/airbyte-ci/connectors/pipelines/pipelines/format/__init__.py#L85

**Comment by bnchrch** [comment]
Related: https://github.com/airbytehq/airbyte/issues/28821

**Comment by alafanechere** [comment]
Definition of done:
On unhandled error we should make sure the dagger logs are always written to disk and uploaded. Should be fixed by removing the early `sys.exit`.

**Comment by alafanechere** [comment]
According to [this failure](https://airbytehq-team.slack.com/archives/C056HGD1QSW/p1690883920523929) the `finally` clause is properly executed and dagger logs are correctly uploaded
